##########################################
Date: Apr 16, 2024 1:00 PM
From: 
Subject: Remove duplicated tags from beacon
Source: Comcast
License: BSD
Upstream-Status: Pending
Signed-off-by: Bogdan Bogush <bogdan.bogush@comcast.com>
##########################################
diff --git a/source/hostap-2.10/src/ap/ap_drv_ops.c b/source/hostap-2.10/src/ap/ap_drv_ops.c
index b0ed432..6df94fd 100644
--- a/source/hostap-2.10/src/ap/ap_drv_ops.c
+++ b/source/hostap-2.10/src/ap/ap_drv_ops.c
@@ -84,7 +84,11 @@ int hostapd_build_ap_extra_ies(struct hostapd_data *hapd,
 		goto fail;
 
 	pos = buf;
+
+/* avoid duplicated extended cap tag in beacon */
+#ifndef CONFIG_DRIVER_BRCM
 	pos = hostapd_eid_ext_capab(hapd, pos);
+#endif
 	if (add_buf_data(&assocresp, buf, pos - buf) < 0)
 		goto fail;
 	pos = hostapd_eid_interworking(hapd, pos);
@@ -112,9 +116,15 @@ int hostapd_build_ap_extra_ies(struct hostapd_data *hapd,
 	if (add_buf_data(&assocresp, buf, pos - buf) < 0)
 		goto fail;
 
+/* avoid duplicated WPS tag in beacon */
+#ifndef CONFIG_DRIVER_BRCM
 	if (add_buf(&beacon, hapd->wps_beacon_ie) < 0 ||
 	    add_buf(&proberesp, hapd->wps_probe_resp_ie) < 0)
 		goto fail;
+#else
+	if (add_buf(&proberesp, hapd->wps_probe_resp_ie) < 0)
+		goto fail;
+#endif
 
 #ifdef CONFIG_P2P
 	if (add_buf(&beacon, hapd->p2p_beacon_ie) < 0 ||
