From 8b3e388551d966083ab2bddccd2b0bd5deba2d61 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Wed, 11 Jan 2023 14:00:44 +0000
Subject: [PATCH] Fix flushing with 0 playback rate

---
 .../platform/graphics/gstreamer/mse/PlaybackPipeline.cpp        | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
index da8ff4b597b3..e4e3ed32bd21 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
@@ -355,6 +355,8 @@ void PlaybackPipeline::flush(AtomString trackId)
 
 #if ENABLE(INSTANT_RATE_CHANGE)
     rate = m_webKitMediaSrc->priv->mediaPlayerPrivate->rate();
+    if (!rate)
+        rate = 1.0;
 #endif
 
     GST_DEBUG_OBJECT(appsrc, "segment: [%" GST_TIME_FORMAT ", %" GST_TIME_FORMAT "], rate: %f",
